### DISCUSSION TITLE
How to push SBOMs to maven-central

### DISCUSSION THREAD

**Discussion by MartinWitt** [jreleaser/jreleaser#1412] [Q&A]
Hey,

I tried to publish SBOMs with our releases, see https://github.com/chains-project/maven-lockfile/blob/main/jreleaser.yml. How can I also push the additional files to maven-central, specially the SBOM? Did I miss some config options?

**Comment by aalmiray** [comment]
The option to deploy all kinds of artifacts to Maven compatible repositories was added in `1.6.0` https://github.com/jreleaser/jreleaser/issues/1135

JReleaser 1.6.0 itself produces SBOMs via the gradle cyclonedx plugin. These artifacts were successfully deployed to MC https://repo1.maven.org/maven2/org/jreleaser/jreleaser/1.6.0/

**Reply by MartinWitt** [reply]
I was searching for the config option, because I have a SBOM produced by cyclonedx-maven-plugin. How is the setting called?

**Reply by aalmiray** [reply]
There's no additional setting required. All staged artifacts found in `stagingRepositories` will be automatically processed. You can confirm this by locally staging artifacts and running `mvn jreleaser:deploy -Djreleaser.dryrun=true`.

**Reply by MartinWitt** [reply]
I run `mvn --no-transfer-progress --batch-mode -Ppublication clean deploy -DaltDeploymentRepository=local::file:./target/staging-deploy`. Afterwards `target/staging-deploy` looks like this:
```
Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
d----          15.06.2023    12:21                io
-a---          15.06.2023    12:21         717077 bom.json
-a---          15.06.2023    12:21         609613 bom.xml
```
Then I invoke jreleaser but the bom.json and bom.xml are afterward not on maven-central, https://repo1.maven.org/maven2/io/github/chains-project/maven-lockfile/3.3.0/.
Attached is the log. [trace.log](https://github.com/jreleaser/jreleaser/files/11756314/trace.log). 
Did I miss an important step?

**Reply by aalmiray** [reply]
Shouldn’t those BOM files be inside the directory matching the groupId as the rest of artifacts? I would venture a guess that’s the reason why they are skipped. Even if they were not skipped, Maven Central would complain due to mismatched file location.

**Comment by LogFlames** [comment]
Hi,
I have taken over the maintanance of maven-lockfile and is trying to get this working.

1. Our release action runs `mvn --no-transfer-progress --batch-mode -Ppublication clean deploy -DaltDeploymentRepository=local::file:./target/staging-deploy` which also trigger the cyclonedx-maven-plugin which generates, attaches and stages them. When running this locally everything works as expected (.json files are placed alongside the .jar's and .pom's in the staging directory). The `tree` of `target` folder is pasted below. I am not worried about bom.json and bom.xml at the moment but trying to get maven-lockfile-5.2.2-cyclonedx.json etc to be uploaded.

<details>
<summary>tree target/</summary>
```
target
├── bom.json
├── bom.xml
└── staging-deploy
    └── io
        └── github
            └── chains-project
                ├── maven-lockfile
                │   ├── 5.2.2
                │   │   ├── maven-lockfile-5.2.2-cyclonedx.json
                │   │   ├── maven-lockfile-5.2.2-cyclonedx.json.md5
                │   │   ├── maven-lockfile-5.2.2-cyclonedx.json.sha1
                │   │   ├── maven-lockfile-5.2.2-cyclonedx.xml
                │   │   ├── maven-lockfile-5.2.2-cyclonedx.xml.md5
                │   │   ├── maven-lockfile-5.2.2-cyclonedx.xml.sha1
                │   │   ├── maven-lockfile-5.2.2-javadoc.jar
                │   │   ├── maven-lockfile-5.2.2-javadoc.jar.md5
                │   │   ├── maven-lockfile-5.2.2-javadoc.jar.sha1
                │   │   ├── maven-lockfile-5.2.2-sources.jar
                │   │   ├── maven-lockfile-5.2.2-sources.jar.md5
                │   │   ├── maven-lockfile-5.2.2-sources.jar.sha1
                │   │   ├── maven-lockfile-5.2.2-tests.jar
                │   │   ├── maven-lockfile-5.2.2-tests.jar.md5
                │   │   ├── maven-lockfile-5.2.2-tests.jar.sha1
                │   │   ├── maven-lockfile-5.2.2.jar
                │   │   ├── maven-lockfile-5.2.2.jar.md5
                │   │   ├── maven-lockfile-5.2.2.jar.sha1
                │   │   ├── maven-lockfile-5.2.2.pom
                │   │   ├── maven-lockfile-5.2.2.pom.md5
                │   │   └── maven-lockfile-5.2.2.pom.sha1
                │   ├── maven-metadata.xml
                │   ├── maven-metadata.xml.md5
                │   └── maven-metadata.xml.sha1
                ├── maven-lockfile-github-action
                │   ├── 5.2.2
                │   │   ├── maven-lockfile-github-action-5.2.2-cyclonedx.json
                │   │   ├── maven-lockfile-github-action-5.2.2-cyclonedx.json.md5
                │   │   ├── maven-lockfile-github-action-5.2.2-cyclonedx.json.sha1
                │   │   ├── maven-lockfile-github-action-5.2.2-cyclonedx.xml
                │   │   ├── maven-lockfile-github-action-5.2.2-cyclonedx.xml.md5
                │   │   ├── maven-lockfile-github-action-5.2.2-cyclonedx.xml.sha1
                │   │   ├── maven-lockfile-github-action-5.2.2-javadoc.jar
                │   │   ├── maven-lockfile-github-action-5.2.2-javadoc.jar.md5
                │   │   ├── maven-lockfile-github-action-5.2.2-javadoc.jar.sha1
                │   │   ├── maven-lockfile-github-action-5.2.2-sources.jar
                │   │   ├── maven-lockfile-github-action-5.2.2-sources.jar.md5
                │   │   ├── maven-lockfile-github-action-5.2.2-sources.jar.sha1
                │   │   ├── maven-lockfile-github-action-5.2.2.jar
                │   │   ├── maven-lockfile-github-action-5.2.2.jar.md5
                │   │   ├── maven-lockfile-github-action-5.2.2.jar.sha1
                │   │   ├── maven-lockfile-github-action-5.2.2.pom
                │   │   ├── maven-lockfile-github-action-5.2.2.pom.md5
                │   │   └── maven-lockfile-github-action-5.2.2.pom.sha1
                │   ├── maven-metadata.xml
                │   ├── maven-metadata.xml.md5
                │   └── maven-metadata.xml.sha1
                └── maven-lockfile-parent
                    ├── 5.2.2
                    │   ├── maven-lockfile-parent-5.2.2-cyclonedx.json
                    │   ├── maven-lockfile-parent-5.2.2-cyclonedx.json.md5
                    │   ├── maven-lockfile-parent-5.2.2-cyclonedx.json.sha1
                    │   ├── maven-lockfile-parent-5.2.2-cyclonedx.xml
                    │   ├── maven-lockfile-parent-5.2.2-cyclonedx.xml.md5
                    │   ├── maven-lockfile-parent-5.2.2-cyclonedx.xml.sha1
                    │   ├── maven-lockfile-parent-5.2.2.pom
                    │   ├── maven-lockfile-parent-5.2.2.pom.md5
                    │   └── maven-lockfile-parent-5.2.2.pom.sha1
                    ├── maven-metadata.xml
                    ├── maven-metadata.xml.md5
                    └── maven-metadata.xml.sha1
```
</details>

2. Next it runs jreleaser, here the sboms (e.g. maven-lockfile-5.2.2-cyclonedx.json) are not included in the `trace.log` as being uploaded. When running `jreleaser full-release --dry-run` locally the cyclonedx.json-files are included.

3. The jreleaser action does upload these files as part of the github release, as specified in `jreleaser.yml` under the `files` section. I cannot, however, find any similar option for the nexus2 deploy.

Would be very thankful for any insight or debugging tips! I cannot figure out what might cause this, as it uploads the files later, "proving" that they were generated by maven during the build step.

**Reply by aalmiray** [reply]
Which version of JReleaser was used?

**Reply by LogFlames** [reply]
Thank you so much! This was the comment I needed to find it.
Such a stupid mistake really, we were using the latest release-action but only JReleaser 1.4.0! That really explains why the behaviour was different both from me running it locally and from reading the source code.

After updating to 1.15.0 everything works as expected!
